zvariant/src/value.rs

```diff
@@ -738,3 +891,163 @@ impl<'a> Type for Value<'a> {
+                ],
+                vec![
+                    Value::new(0_u8),
+                    Value::new((Value::new(51), Value::new(Value::new(1_u32)))),
+                ]
+            ))
+            .to_string(),
+            "([signature '', '(ysa{sd})'], [objectpath '/', '/a/very/looooooooooooooooooooooooo0000o0ng/path'], [<byte 0x00>, <(<51>, <<uint32 1>>)>])"
+        );
+
+        assert_eq!(Value::new(vec![] as Vec<Vec<i64>>).to_string(), "@aax []");
+        assert_eq!(
+            Value::new(vec![
+                vec![0_i16, 1_i16],
+                vec![2_i16, 3_i16],
+                vec![4_i16, 5_i16]
+            ])
+            .to_string(),
+            "[[int16 0, 1], [2, 3], [4, 5]]"
+        );
+        assert_eq!(
+            Value::new(vec![
+                b"Hello".to_vec(),
+                b"Hell\0o".to_vec(),
+                b"Hello\0".to_vec(),
+                b"\0".to_vec(),
+                b" \0".to_vec(),
+                b"'\0".to_vec(),
+                b"\n'\"\0".to_vec(),
+                b"\\\0".to_vec(),
+            ])
+            .to_string(),
+            r#"[[byte 0x48, 0x65, 0x6c, 0x6c, 0x6f], [0x48, 0x65, 0x6c, 0x6c, 0x00, 0x6f], b'Hello', b'', b' ', b"'", b"\n'\"", b'\\']"#
+        );
+
+        assert_eq!(
+            Value::new(HashMap::<bool, bool>::new()).to_string(),
+            "@a{bb} {}"
+        );
+        assert_eq!(
+            Value::new(vec![(true, 0_i64)].into_iter().collect::<HashMap<_, _>>()).to_string(),
+            "{true: int64 0}",
+        );
+        // The order of the entries may vary
+        let val = Value::new(
+            vec![(32_u16, 64_i64), (100_u16, 200_i64)]
+                .into_iter()
+                .collect::<HashMap<_, _>>(),
+        )
+        .to_string();
+        assert!(val.starts_with('{'));
+        assert!(val.ends_with('}'));
+        assert_eq!(val.matches("uint16").count(), 1);
+        assert_eq!(val.matches("int64").count(), 1);
+
+        let items_str = val.split(", ").collect::<Vec<_>>();
+        assert_eq!(items_str.len(), 2);
+        assert!(items_str
+            .iter()
+            .any(|str| str.contains("32") && str.contains(": ") && str.contains("64")));
+        assert!(items_str
+            .iter()
+            .any(|str| str.contains("100") && str.contains(": ") && str.contains("200")));
+
+        assert_eq!(Value::new(Structure::default()).to_string(), "()");
+        assert_eq!(
+            Value::new(((true,), (true, false), (true, true, false))).to_string(),
+            "((true,), (true, false), (true, true, false))"
+        );
+
+        assert_eq!(
+            Value::new((
+                (Some(0_i16), Some(Some(0_i16)), Some(Some(Some(0_i16))),),
+                (None::<i16>, Some(None::<i16>), Some(Some(None::<i16>)),),
+            ))
+            .to_string(),
+            "((@mn 0, @mmn 0, @mmmn 0), (@mn nothing, @mmn just nothing, @mmmn just just nothing))"
```

Comment:
* `just just nothing` seems weird. I think it's easier to just (no pun intended) output `nothing`. i-e if the outer maybe is `None` than that's all the value that you need to know about (especially since we're printing the signature).
* wouldn't `0@mn` be better? It'd also be consistent with Rust's way of adding the type at the end. Also, maybe we just use the Rust syntax? This is afterall a Rust API and most users know Rust a lot more than D-Bus. We can leave the signature then to the `Debug` output.

Reply:
> * `just just nothing` seems weird. I think it's easier to just (no pun intended) output `nothing`. i-e if the outer maybe is `None` than that's all the value that you need to know about (especially since we're printing the signature).

I think it would cause a problem if you have something like `Some(None::<Option<i16>>)`. If you only have `nothing` it won't be possible to know where the `nothing` begins unlike in `@mmmn just nothing`.

>  * wouldn't `0@mn` be better? It'd also be consistent with Rust's way of adding the type at the end. Also, maybe we just use the Rust syntax? This is afterall a Rust API and most users know Rust a lot more than D-Bus. We can leave the signature then to the `Debug` output.

I don't have a strong opinion, but I was trying to be consistent with `GVariant` printing.

Reply:
> I think it would cause a problem if you have something like Some(None::<Option<i16>>). If you only have nothing it won't be possible to know where the nothing begins unlike in @mmmn just nothing.

Hmm.. true.

> I don't have a strong opinion, but I was trying to be consistent with GVariant printing.

Ah, I didn't realize there was prior art. In that case, just follow that. The bonus would be that we don't need to hardcode the strings in gvariant format tests but rather compare the glib's (it's a dev dep) output.

BTW, I didn't check but please do make sure that all  gvariant feature requiring code is under a `cfg` (didn't check if it is already or not, just a reminder).

Reply:
> BTW, I didn't check but please do make sure that all gvariant feature requiring code is under a cfg (didn't check if it is already or not, just a reminder).

I did check by testing it with the `gvariant` feature disabled. Not sure if that was enough and, hopefully, I didn't miss a `cfg`.

Reply:
Cool. As I wrote, I didn't mean to imply that you missed anything. Just came to my head when I was writing my comment so expressed the thought before I forget it completely. :laughing: